db/kv_dense_channel_storage_reader.go

```diff
@@ -0,0 +1,221 @@
+		sinceSeq := sinceClock.GetSequence(vbNo)
+		if sinceSeq < toSeq {
+			changedVbs = append(changedVbs, vbNo)
+			partitionNo := ds.partitions.PartitionForVb(vbNo)
+			if changedPartitions[partitionNo] == nil {
+				partitionRange := NewPartitionRange()
+				changedPartitions[partitionNo] = &partitionRange
+			}
+			changedPartitions[partitionNo].SetRange(vbNo, sinceSeq, toSeq)
+		}
+	}
+	return changedVbs, changedPartitions
+}
+
+// DensePartitionStorageReader is a non-caching reader - every read request retrieves the latest from the bucket.
+type DensePartitionStorageReader struct {
+	channelName string      // Channel name
+	partitionNo uint16      // Partition number
+	indexBucket base.Bucket // Index bucket
+}
+
+type PartitionChanges struct {
+	changes map[uint16][]*LogEntry
+}
+
+func NewPartitionChanges() *PartitionChanges {
+	return &PartitionChanges{
+		changes: make(map[uint16][]*LogEntry),
+	}
+}
+func (p *PartitionChanges) GetVbChanges(vbNo uint16) []*LogEntry {
+	if vbchanges, ok := p.changes[vbNo]; ok {
+		return vbchanges
+	} else {
+		return nil
+	}
+}
+
+func (p *PartitionChanges) AddEntry(entry *LogEntry) {
+	_, ok := p.changes[entry.VbNo]
+	if !ok {
+		p.changes[entry.VbNo] = make([]*LogEntry, 0)
+	}
+	p.changes[entry.VbNo] = append(p.changes[entry.VbNo], entry)
+}
+
+func (p *PartitionChanges) Count() int {
+	count := 0
+	for _, vbSet := range p.changes {
+		count += len(vbSet)
+	}
+	return count
+}
+
+func NewDensePartitionStorageReader(channelName string, partitionNo uint16, indexBucket base.Bucket) *DensePartitionStorageReader {
+	storage := &DensePartitionStorageReader{
+		channelName: channelName,
+		partitionNo: partitionNo,
+		indexBucket: indexBucket,
+	}
+	return storage
+}
+
+func (r *DensePartitionStorageReader) GetChanges(partitionRange PartitionRange) (*PartitionChanges, error) {
+
+	changes := NewPartitionChanges()
+
+	// Initialize the block list to the starting range, then find the starting block for the partition range
+	blockList := r.GetBlockListForRange(partitionRange)
+	startIndex := 0
+	for startIndex < len(blockList.blocks) {
+		if partitionRange.SinceAfter(blockList.blocks[startIndex].StartClock) {
+			startIndex++
+		} else {
+			// We've gone too far - want to start with the previous block
+			break
+		}
+	}
+	startIndex--
+
+	// Iterate over the blocks, collecting entries by vbucket
+	for i := startIndex; i <= len(blockList.blocks)-1; i++ {
+		blockIter := NewDenseBlockIterator(blockList.LoadBlock(blockList.blocks[i]))
+		for {
+			logEntry := blockIter.next()
+			if logEntry == nil {
+				break
+			}
+			switch compare := partitionRange.Compare(logEntry.VbNo, logEntry.Sequence); compare {
+			case PartitionRangeAfter:
+				break // We've exceeded the range - return
+			case PartitionRangeWithin:
+				changes.AddEntry(logEntry)
+			}
+		}
+	}
+	return changes, nil
+}
+
+func (r *DensePartitionStorageReader) GetBlockListForRange(partitionRange PartitionRange) *DenseBlockList {
```

Comment:
Implicitly tested by the GetChanges test.  In particular TestChannelStorageCorrectness_Dense in sg_accel writes enough data to span multiple blocks, and so verifies the BlockListForRange functionality.